distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso

```diff
@@ -572,14 +573,41 @@ type Vector a
        Arguments:
        - function: A function that takes an element in the vector and returns
          some transformation of that element.
+       - on_problems: Specifies how to handle any problems that arise in
+         `function`.
+
+       ! Error Conditions
+
+         Errors and Warnings that arise when executing the function are wrapped
+         in `Map_Error`.
+
+         Only `MAX_MAP_WARNINGS` warnings are attached to result values. After
+         that, the warnings are dropped, but a count of the additional warnings
+         is attached in an `Additional_Warnings` warning.
+
+       ? Problem Handling
+
+         The result of Errors thrown when executing the function depend on `on_problems`:
+         - Problem_Behavior.Report_Error: The error is thrown.
+         - Problem_Behavior.Report_Warning: The result is `Nothing`, with the
+           error attached as a warning.
+         - Problem_Behavior.Ignore: The result is `Nothing`, and the error
+           attached as a warning.
 
        > Example
          Add 1 to each element of the vector.
 
              [1, 2, 3] . map +1
-    map : (Any -> Any) -> Vector Any
-    map self function =
-        Vector.new self.length i-> function (self.at i)
+    map : (Any -> Any) -> Problem_Behavior -> Vector Any
+    map self function on_problems=Problem_Behavior.Report_Error =
+        Array_Like_Helpers.vector_from_function self.length (function << self.at) on_problems
+
+    ## PRIVATE
+       Implementation of `map` that uses `vector_from_function_primitive`, which
+       does not do error wrapping, and does not have `on_problems`.
+    map_primitive : (Any -> Any) -> Vector Any
```

Comment:
I'd like to avoid these helper private methods in the `Vector` and `Array` & other public classes of (not only) builtins. 

Please move them into `Array_Like_Helper` and make them extension methods. Mark `Array_Like_Helper` as `private`, so nobody can access it outside of the `Standard.Base` project.

Let me know if something doesn't work - I can help to make this dream done.

Reply:
I think we may want to use these methods _outside_ of the `Base` module as well.

They are not 'private helpers' but just a 'developer oriented' variant of `map` that is more oriented towards internal implementations than the user-facing map whose error handling is more IDE-centric.

So I don't think they should necessarily be that hidden - they are not meant for the GUI users primarily, but library devs on outside libraries will definitely appreciate this function.

---

Although I agree the name may be misleading, maybe we should come up with a better one - although I don't know what could work better here. IMHO the current name is okay, but maybe we can come up with one that makes the intent clearer? 

I guess I'd firstly change the doc comment - it does not need to inform about implementation details like `vector_from_function_primitive` that are not clear for users/developers that don't know it. IMO it should talk functionality - it's a variant of `map` that is closer to classical programming languages - it fails at first error and returns it as-is; as opposed to our user-centric `map` which does error wrapping and recovery depending on settings.

Reply:
@JaroslavTulach Indeed these are used in Standard/Table, out of necessity. Without using them, errors are wrapped in `Map_Error` because the internal implementation uses `map` in places.

I'm not very happy with having to sprinkle calls to map_primitive around the code to remove clutter from user-facing errors, I think it might be an ongoing chore. Alternately, we might find that we just use the `*_primitive` forms by default in all library code, which is awkward but perhaps necessary. The underlying problem is that `Vector/map` are fundamental to functional programming, and library implementors and end-users have different needs -- even though they are sometimes the same person.

Perhaps what we ultimately need is a user-facing `Vector` that mostly delegates to the internal `Vector`, but also implements things like `Map_Error` wrapping.

Reply:
Another example of complexity that comes up here is that `vector_from_function` needs to manipulate errors and warnings. If any code involved in that manipulation itself uses `map` there's the possibility of a bottomless recursion -- as happened to me several times in implementing this.

Reply:
> Perhaps what we ultimately need is a user-facing `Vector` that mostly delegates to the internal `Vector`, but also implements things like `Map_Error` wrapping.

That seems _really_ risky. The probability of it leaking to the user and confusing them (why the API behaves differently with this instance than the other time??) is IMO far too great.

I think given that we have different needs of different users (GUI vs libs) we just need to stick to using these different variants... We could merge it into one and add `on_problems=Some_Special_Value` to tell it to use the 'primitive' mode, but IMHO that is even more brittle - the `map_primitive` seems to be a solution that sounds rather simple and robust to me.

Reply:
There's no risk if we have a way to prevent undocumented classes from being exposed to the user, which we definitely need. I don't suggest we call them both Vector -- they're different this, but share an implementation.

For now I'll keep these private and update the docs.

Reply:
> There's no risk if we have a way to prevent undocumented classes from being exposed to the user, which we definitely need.

Interesting proposal, but I don't think we will be able to get this kind of functionality very soon.

Reply:
> > There's no risk if we have a way to prevent undocumented classes from being exposed to the user, which we definitely need.
> 
> Interesting proposal, but I don't think we will be able to get this kind of functionality very soon.

FYI: Tracked in https://github.com/enso-org/enso/issues/8100

Reply:
> > > There's no risk if we have a way to prevent undocumented classes from being exposed to the user, which we definitely need.
> > 
> > 
> > Interesting proposal, but I don't think we will be able to get this kind of functionality very soon.
> 
> FYI: Tracked in #8100

That does not seem to be related?

The ticket is about private fields of an object. The thing discussed here is about some 'types' not being able to leak into user-facing code. Not the content of the type, but the type itself.

I assume it would need to be based on return-value checks.

Reply:
> They are not 'private helpers' but just a 'developer oriented' variant of `map` that is more oriented towards internal implementations than the user-facing map whose error handling is more IDE-centric.
> 
> So I don't think they should necessarily be that hidden - they are not meant for the GUI users primarily, but library devs on outside libraries will definitely appreciate this function.

Is the above said by the same @radeusgd that usually takes _rationalistic point of view_? This is not a rationalistic approach at all. This stance is horribly _pragmatic_.

However we are talking about _core API class of Enso_. Where else shall we act _rationalisticly_ than here? Moreover we are talking about builtin - e.g. _part of the engine_. Hence, I am willing to fight against debris like this to appear in these _core classes_.

> Perhaps what we ultimately need is a user-facing Vector that mostly delegates to the internal Vector, but also implements things like Map_Error wrapping.

If you want internal array like object while implementing a library, use `java.util.ArrayList` or co. We haven't 
- #6218

to introduce two types of vectors again.

> prevent undocumented classes from being exposed to the user

Yes, please use `private` modules (e.g. real _private_, not just a comment). Should any problem (like #8469) appear, speak up.